// @directus/constants/dist/index.js
// src/activity.ts
var Action = /* @__PURE__ */ ((Action2) => {
  Action2["CREATE"] = "create";
  Action2["UPDATE"] = "update";
  Action2["DELETE"] = "delete";
  Action2["REVERT"] = "revert";
  Action2["VERSION_SAVE"] = "version_save";
  Action2["COMMENT"] = "comment";
  Action2["UPLOAD"] = "upload";
  Action2["LOGIN"] = "login";
  Action2["RUN"] = "run";
  return Action2;
})(Action || {});

// src/fields.ts
var KNEX_TYPES = [
  "bigInteger",
  "boolean",
  "date",
  "dateTime",
  "decimal",
  "float",
  "integer",
  "json",
  "string",
  "text",
  "time",
  "timestamp",
  "binary",
  "uuid"
];
var TYPES = [
  ...KNEX_TYPES,
  "alias",
  "hash",
  "csv",
  "geometry",
  "geometry.Point",
  "geometry.LineString",
  "geometry.Polygon",
  "geometry.MultiPoint",
  "geometry.MultiLineString",
  "geometry.MultiPolygon",
  "unknown"
];
var GEOMETRY_TYPES = [
  "Point",
  "LineString",
  "Polygon",
  "MultiPoint",
  "MultiLineString",
  "MultiPolygon"
];
var GEOMETRY_FORMATS = ["native", "geojson", "wkt", "lnglat"];
var LOCAL_TYPES = [
  "standard",
  "file",
  "files",
  "m2o",
  "o2m",
  "m2m",
  "m2a",
  "presentation",
  "translations",
  "group"
];
var RELATIONAL_TYPES = [
  "file",
  "files",
  "m2o",
  "o2m",
  "m2m",
  "m2a",
  "presentation",
  "translations",
  "group"
];
var FUNCTIONS = ["year", "month", "week", "day", "weekday", "hour", "minute", "second", "count"];

// src/files.ts
var JAVASCRIPT_FILE_EXTS = ["js", "mjs", "cjs"];

// src/injection.ts
var STORES_INJECT = "stores";
var API_INJECT = "api";
var EXTENSIONS_INJECT = "extensions";

// src/regex.ts
var REGEX_BETWEEN_PARENS = /\(([^)]+)\)/;
export {
  API_INJECT,
  Action,
  EXTENSIONS_INJECT,
  FUNCTIONS,
  GEOMETRY_FORMATS,
  GEOMETRY_TYPES,
  JAVASCRIPT_FILE_EXTS,
  KNEX_TYPES,
  LOCAL_TYPES,
  REGEX_BETWEEN_PARENS,
  RELATIONAL_TYPES,
  STORES_INJECT,
  TYPES
};
